### ISSUE TITLE
Refine support for Intel compiler and more optimization flags

### ISSUE THREAD

**Issue by weijianwen** [dmlc/dmlc-core#28]
In many cases, simply recompiling code against Intel compiler (ICC) and Math Kernel Library (MKL) gains 5%~15% performance improvement over GCC. So I propose to give users the choice to build dmlc-core with ICC or GCC. 

Wider vectorization has been introduced in recent (?Intel only?) CPU. For example, AVX with 256-bit SIMD (for float only) since Sandy Bridge in 2012 and AVX2 with 256-bit SIMD (for both float and integer) since Haswell in 2014. Compiler options can be introduced to enable AVX or AVX2, for example ```USE_AVX``` and ```USE_AVX2```.

Even the dmlc-core's building process is straight-forward enough, adding sections like "Build" and "Install" in README will pack all the necessary things together. This also makes dmlc-core more like a standalone deployable package.Cluster administrators would love to see that.

In summary,

1. Add support for Intel compiler.
2. Add support for AVX, AVX2.
3. Add ```Build``` and ```Install``` sections in README.

I think I can help and run some tests on a standard cluster. But that would be after mid August.

**Comment by mli** [comment]
i think it can be done by changing CXX in config.mk into `icc -march=core-avx2` or simply overwrite CXX by `make CXX="icc -march=core-avx2"`

**Comment by weijianwen** [comment]
To take AVX as an example, simply change ```CXX``` to ```icpc -xavx``` in ```make/config``` doesn't work. The advanced vectorization flag (avx) will be override by the old vectorization flag (sse2) which is introduced by ```CFLAGS```in ```Makefile```.

```
icpc -xavx -c -O3 -Wall -msse2  -Wno-unknown-pragmas -Iinclude -std=c++0x -fopenmp -fPIC -DDMLC_USE_HDFS=0 -DDMLC_USE_S3=0 -o line_split.o src/io/line_split.cc
icpc: command line warning #10120: overriding '-xavx' with '-msse2
```

Given that different compilers may have (slightly?) different compiling and linking flags, Specification for compiler, CFLAGS and LDFLAGS should be packed togehter. I propose:

1. Set up different ```config.mk``` (config.mk.gcc, config.mk.icc, config.mk.pgi...) for different compilers.
2. Move ```CFLAGS``` and ```LDFLAGS``` to ```config.mk``` where customization can be made in one file.
3. Regarding Intel Compiler, ```-xhost``` is favored over specific vectorization flags such as ```-msse2``` or ```-xavx```. ```-xhost``` will automatically *generate instruction sets up to the highest that is supported by the compilation host* (for both Intel and Intel-compatible host).

Here is the snip of ```config.mk``` for Intel compiler:

```
export CC = icc -xhost
export CXX = icpc -xhost
export MPICXX = mpiicpc
export LDFLAGS= -pthread -lm
export CFLAGS = -O3 -Wall -Wno-unknown-pragmas -Iinclude -std=c++0x -ip
```

**Comment by weijianwen** [comment]
I will also try to build dmlc-core with ```-march=native``` in GCC. Hopes it won't break things. There should be NO compatibility issue if the building host and running host use CPUs with the same HW features. I think it is common practice in cluster.

I am not sure whether it is worthy to change the default flag from ```-msse2``` to ```-march=native```, in trade of some performance gain (if any). What do you guys think?

**Comment by weijianwen** [comment]
I'll close this issue and open two PR after my verication:

1. Move "CFLAGS" adn "LDFLAGS" to ```make/config```. Replace "-msse2" with "-march=native".
2. Add ```make/config.icc``` for Intel Compiler.

Best,
